### PR TITLE
feat(ADE-495): Add base blueprint pattern

### PR DESCRIPTION
**Pull request by SamMurphyDev**

### CODE REVIEW

File: idea_search.iml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module org.jetbrains.idea.maven.project.MavenProjectsManager.isMavenModule="true" type="JAVA_MODULE" version="4">
```
Comment (robinMcA) [inline idea_search.iml:2]: This is in the ignore file, delete?

File: src/main/resources/templates/blueprint.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<h1>
+    <at:var at:name="vIdeaTitle" at:rawxhtml="true"/>
```
Comment (robinMcA) [inline src/main/resources/templates/blueprint.xml:2]: I don't think that we need this title anymore

File: src/main/resources/soy/ideaSearch.soy [hunk 3]
```diff
@@ -0,0 +1,47 @@
+{namespace MyPlugin.Blueprints.Simple}
```
Comment (robinMcA) [inline src/main/resources/soy/ideaSearch.soy:1]: Should this name change?

File: src/main/resources/js/ideaSearch.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  'au.com.agiledigital.idea_search:create-by-sample-template',
+  function (wizard) {
+    wizard.on('submit.page1Id', function (e, state) {
+      console.log(state.pageData);
+      var ideaTitle = state.pageData.vIdeaTitle;
+      var ideaDescription = state.pageData.vIdeaDescription;
+      var ideaOwner = state.pageData.vIdeaOwner;
+      var ideaStatus = state.pageData.vIdeaStatus;
+      var ideaTeam = state.pageData.vIdeaTeam;
+      console.log(ideaDescription, ideaOwner, ideaStatus, ideaTeam);
```
Comment (robinMcA) [inline src/main/resources/js/ideaSearch.js:11]: can get rid of these

File: src/main/java/au/com/agiledigital/idea_search/Status.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package au.com.agiledigital.idea_search;
```
Comment (robinMcA) [inline src/main/java/au/com/agiledigital/idea_search/Status.java:1]: ToDo: Task to create a model that has this and the idea in it?

File: src/main/java/au/com/agiledigital/idea_search/Options.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+public class Options {
+
+  public boolean isDefault = false;
+  public boolean isUser = false;
+  public boolean isStatus = false;
+
+  /**
+   * Sets the flag for default value
+   *
+   * @param aDefault Is a default value
```
Comment (robinMcA) [inline src/main/java/au/com/agiledigital/idea_search/Options.java:12]: nit: aDefault ?

File: src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  @Override
+  protected BlueprintContext updateBlueprintContext(
+    BlueprintContext blueprintContext
+  ) {
+    Map<String, Object> contextMap = blueprintContext.getMap();
+
+    blueprintContext.setTitle(contextMap.get("vIdeaTitle").toString());
+
+    contextMap
```
Comment (robinMcA) [inline src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java:133]: A comment for this?

File: src/main/java/au/com/agiledigital/idea_search/Options.java (hunk 6 above)
Comment (SamMurphyDev) [reply src/main/java/au/com/agiledigital/idea_search/Options.java:12]: `default` is a reversed keyword in java. I could change to `isDefault`

File: src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java (hunk 7 above)
Comment (SamMurphyDev) [reply src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java:132]: Added comment

File: idea_search.iml (hunk 1 above)
Comment (SamMurphyDev) [reply idea_search.iml]: Add to .gitignore

File: src/main/resources/templates/blueprint.xml (hunk 2 above)
Comment (SamMurphyDev) [reply src/main/resources/templates/blueprint.xml]: Removed from template

File: src/main/resources/soy/ideaSearch.soy (hunk 3 above)
Comment (SamMurphyDev) [reply src/main/resources/soy/ideaSearch.soy]: Changed to use `IdeaSearch.Blueprints.Simple` namespace

File: src/main/resources/js/ideaSearch.js (hunk 4 above)
Comment (SamMurphyDev) [reply src/main/resources/js/ideaSearch.js]: Removed logging

File: src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    Map<String, Object> contextMap = blueprintContext.getMap();
+
+    blueprintContext.setTitle(contextMap.get("vIdeaTitle").toString());
+
+    // Goes through map and adds in default values and transforms pre-existing values
+    contextMap
+      .entrySet()
+      .forEach(
+        entry ->
+          ifElseComputeLossyFalse(
```
Comment (dbalmain) [inline src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java:136]: This seems overly complex to me. Is `ifElseComputeLossyFalse` going to get used a lot in this function? I think you could simplify it using https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#compute-K-java.util.function.BiFunction-:

```java
contextMap.compute((k, v) -> v == null || (value instanceof String && ((String) value).length() == 0)
  ? defaults
￼     .stream()
￼     .filter(property -> property.key.equals(k))
￼     .findFirst()
￼     .orElse(
￼       new KeyProperty(
￼         k,
￼         "Something went very wrong here",
￼         setupOptions(k,
￼           new Options().withDefault(true))
￼       )
￼     )
  : new KeyProperty(
      k,
      entry.getValue(),
      setupOptions(k,
        new Options().withDefault(
          false))
    )
```

File: src/main/resources/soy/ideaSearch.soy [hunk 9]
```diff
@@ -0,0 +1,48 @@
+{namespace IdeaSearch.Blueprints.FedexIdea}
+
+/**
+*  A form that accepts a person's name
```
Comment (dbalmain) [inline src/main/resources/soy/ideaSearch.soy:4]: Stale comment

Comment (SamMurphyDev) [reply src/main/resources/soy/ideaSearch.soy]: Updated commet

File: src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java (hunk 8 above)
Comment (SamMurphyDev) [reply src/main/java/au/com/agiledigital/idea_search/BlueprintContextProvider.java]: Changed to BiFunction and removed redundant code

File: README.md [hunk 10]
```diff
@@ -12,8 +12,15 @@ Either add the addon throught he Atlassian Market place (link to come) or by dir
...
 
-- Clone the repository, install the Atlassian SDK,
-- In the root directory run atlas-run
+1. Clone the repository
+2. Install the [Atlassian SDK](https://developer.atlassian.com/server/framework/atlassian-sdk/)
+3. In the root directory run `atlas-run`
+4. Once Confluence has started you can access it via http://localhost:1990/confluence. Note you will need to change the baseUrl in the confluence settings otherwise you will have issues with the creation of resources
+
+- To run without tests use the command `atlas-run -Dmaven.test.skip=true`
+- We suggest dumpings logs to a file due to the large amount lines that are outputted. A command to do this would `rm -rf logs.log; atlas-run > logs.log`
```
[the remark this answers is not in the document]
Comment (SamMurphyDev) [reply README.md]: Changed to use suggestion

File: README.md [hunk 11]
```diff
@@ -12,8 +12,15 @@ Either add the addon throught he Atlassian Market place (link to come) or by dir
...
+
+To run the hot reload, have the confluence server running. Run the command `atlas-mvn package` which will build the app. The server will see the change and hot load the new version into the app.
```
[the remark this answers is not in the document]
Comment (SamMurphyDev) [reply README.md:23]: Changed wording to reflect comment

**Review by dbalmain — APPROVED** [review]
Nice work Sam!

**Pull request merged**